File: extensions/aliexpress/src/test/java/app/morphe/extension/aliexpress/patches/DisableAnalyticsPatchTest.java
package app.morphe.extension.aliexpress.patches;

import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertFalse;

public class DisableAnalyticsPatchTest {

    @Test
    public void testShouldTrackEvent() {
        assertFalse(DisableAnalyticsPatch.shouldTrackEvent());
    }

    @Test
    public void testRewriteAnalyticsUrl_WithNull() {
        assertNull("Null URL should return null", DisableAnalyticsPatch.rewriteAnalyticsUrl(null));
    }

    @Test
    public void testRewriteAnalyticsUrl_WithEmptyString() {
        assertEquals("Empty string should be returned as-is", "", DisableAnalyticsPatch.rewriteAnalyticsUrl(""));
    }

    @Test
    public void testRewriteAnalyticsUrl_WithBenignUrls() {
        String[] benignUrls = {
            "https://aliexpress.com/api/products",
            "https://gw.alicdn.com/imgextra/i3/O1CN01...",
            "https://login.aliexpress.com",
            "/local/cache/images"
        };

        for (String url : benignUrls) {
            assertEquals("Benign URL should not be blocked", url, DisableAnalyticsPatch.rewriteAnalyticsUrl(url));
        }
    }

    @Test
    public void testRewriteAnalyticsUrl_WithTrackingUrls() {
        String[] trackingUrls = {
            "https://acs.aliexpress.com/gw/mtop.aliexpress.analytics.collect/1.0/",
            "https://log.aliexpress.com/track/event",
            "https://umeng.com/api/v1/collect",
            "https://sensorsdata.aliexpress.com",
            "https://ump.aliexpress.com/report"
        };

        for (String url : trackingUrls) {
            assertNull("Tracking URL should be blocked (returned as null)", DisableAnalyticsPatch.rewriteAnalyticsUrl(url));
        }
    }
}

File: patches/src/test/kotlin/app/morphe/patches/aliexpress/analytics/DisableAnalyticsPatchTest.kt
package app.morphe.patches.aliexpress.analytics

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Test

class DisableAnalyticsPatchTest {

    @Test
    fun verifyAnalyticsInitFingerprintBoundaries() {
        val fingerprint = AnalyticsInitFingerprint

        assertEquals("V", fingerprint.returnType)
        assertNotNull(fingerprint.strings)
        assertEquals(2, fingerprint.strings?.size)
        assertTrue(fingerprint.strings?.contains("setGlobalProperty") ?: false)
        assertTrue(fingerprint.strings?.contains("key is null or key is empty or value is null,please check it!") ?: false)
    }

    @Test
    fun verifyDisableAnalyticsPatchProperties() {
        val patch = disableAnalyticsPatch

        assertEquals("Disable analytics", patch.name)
        assertEquals("Blocks analytics and tracking sent to AliExpress servers.", patch.description)
        assertTrue(patch.default)
    }
}
